doc/sonic-vrf-hld-v0.2.md

```diff
@@ -0,0 +1,412 @@
+
+“mtu”:”1500”
+
+“vrf”:”vrf-blue”}
+
+“Ethernet1”:{
+
+“mtu”:”1500”
+
+“vrf”:”vrf-red”}
+
+"Ethernet0\|11.11.11.1/24": {},
+
+"Ethernet1\|12.12.12.1/24": {},
+
+"Ethernet2\|13.13.13.1/24": {},
+
+"Ethernet3\|14.14.14.1/24": {},
+
+},
+
+"VLAN_INTERFACE": {
+
+“Vlan100”:{
+
+“mtu”:”1500”
+
+“vrf”:”vrf-blue”}
+
+"Vlan100\|15.15.15.1/24": {},
+
+}
+
+1.  **Adding a app-VRF-table**
+
+;defines virtual routing forward table
+
+;
+
+;Status: stable
+
+key = VRF_TABLE:VRF_NAME ;
+
+fall_through = “1”/”0”
+
+1.  **Breaking up app-intf-table into app-intf-table and app-intf-prefix-table**
+
+app-intf-table is defined as the following:
+
+;defines logical network interfaces, an attachment to a PORT name
+
+;
+
+;Status: stable
+
+key = INTF_TABLE:ifname
+
+if_mtu = 1\*4DIGIT ; MTU for the interface
+
+VRF_NAME = 1\*64VCHAR ;
+
+app-intf-prefix-table is defined as the following:
+
+;defines logical network interfaces with IP-prefix, an attachment to a PORT and
+list of 0 or more ip prefixes;
+
+;Status: stable
+
+key = INTF_TABLE:ifname:IPprefix ; an instance of this key will be repeated for
+each prefix
+
+IPprefix = IPv4prefix / IPv6prefix ; an instance of this key/value pair will be
+repeated for each prefix
+
+scope = "global" / "local" ; local is an interface visible on this localhost
+only
+
+if_mtu = 1\*4DIGIT ; MTU for the interface （move to INTF_TABLE:ifname table）
+
+family = "IPv4" / "IPv6" ; address family
+
+1.  **Adding VRF key to app-route-table key list**
+
+;Stores a list of routes
+
+;Status: Mandatory
+
+key = ROUTE_TABLE:VRF_NAME:prefix ;
+
+nexthop = \*prefix, ;IP addresses separated “,” (empty indicates no gateway)
+
+intf = ifindex? PORT_TABLE.key ; zero or more separated by “,” (zero indicates
+no interface)
+
+blackhole = BIT ; Set to 1 if this route is a blackhole (or null0)
+
+Since global vrf name is null, global vrf key will becomes ROUTE_TABLE:prefix.
+
+ VrfMgrd 
+---------
+
+>   Listening VRF related configuration in config_db such as VRF
+>   creation/deletion, VRF binding to any interface. Once detected, update
+>   kernel using iproute2 CLIs and write VRF information to app-VRF-table and
+>   app-intf-table.
+
+>   VrfMgrd process will be placed in swss docker. In case of “swss restart”,
+>   VRF device will be still retained in kernel. When VrfMgrd starts up it
+>   querys all master device from kernel and clean up all vrf related device and
+>   restore vrf device per configdb.
+
+>   VRFOrch has three member routerCnt , neighCnt and rifCnt which record VRF
+>   related route number , neigh number and rif number.VRFOrch is added to
+>   RouteOrch , NeighOrch and intfsOrch member to update routeCnt , neighCnt and
+>   rifCnt.
+
+>   When  VRFOrch receives vrf-delete event VRF object won’t  be deleted until
+>   routerCnt  ,neighCnt and rifCnt  is decreased to zero.
+
+>   When device binds to specified VRF, the ip address of slave device will be
+>   removed and kernel will delete all neigh associated slave device.
```

Comment:
I believe on enslavement, connected and local routes are automatically moved to the table associated with the VRF device. Is that not the case with the latest kernel? Can you explain how you plan to achieve the above removal/behavior you mention?